include: riscv: add RISC-V arch atomic operations

achech opened the pull request:
Add RISC-V architecture specific atomic instructions.

Atomic instructions for RISC-V are specified in standard document
`The RISC-V Instruction Set Manual Volume I: Unprivileged ISA`

npitre commented:
Please add some explanation in the commit log explaining why this
is needed and how it differs from using atomic_builtin.h.

And to be effective you need CONFIG_ATOMIC_OPERATIONS_ARCH to be set which
is not part of your patch.

andyross commented:
So... does the toolchain get this wrong?  Or is there a toolchain without support?  What's the advantage here vs. CONFIG_ATOMIC_OPERATIONS_BUILTIN?  Generally my gut tends to trust toolchain authors to keep lockless stuff correct per specification than I do Zephyr people to maintain it.

achech commented:
> Please add some explanation in the commit log explaining why this is needed and how it differs from using atomic_builtin.h.

The atomic operations currently not supported in atomic_builtin.h are `atomic_max/atomic_min`, which RISCV standard defines.
I have tried to update the ~/tests/kernel/common/src/atomic.c with tests for max/min but twister fails to compile them.
Also, the `atomic_swap`, doesn't exist as such, in atomic_builtin.h

> And to be effective you need CONFIG_ATOMIC_OPERATIONS_ARCH to be set which is not part of your patch.

True. To actually compile the added file, ATOMIC_OPERATIONS_ARH must be activated. As far as I was able to see, it
was set in arch/Kconfig for XTENSA arch, with
       ` imply ATOMIC_OPERATIONS_ARCH`
but was later removed. So, the activation should go through ~/soc/.../Kconfig.soc, I think.

achech commented:
> So... does the toolchain get this wrong? Or is there a toolchain without support? What's the advantage here vs. CONFIG_ATOMIC_OPERATIONS_BUILTIN? Generally my gut tends to trust toolchain authors to keep lockless stuff correct per specification than I do Zephyr people to maintain it.

Basically, atomics not currently included in atomic_builtin.h are atomic_max, atomic_min, atomic_swap.
Also, the RISC-V standard "advises" LR/SC as the preferable way to implement CAS.

andyross reviewed (changes requested):
That's because atomic_{min,max,swap}() are not Zephyr APIs, as they don't correspond to common hardware capabilities. 

I really don't think we want this in this form.  For the Zephyr atomics API, we have toolchain support already on all targets (which is the ideal situation, absent known toolchain bugs or misfeatures).    If you want to add a riscv-specific atomics API, that's obviously fine.

achech commented:
> That's because atomic_{min,max,swap}() are not Zephyr APIs, as they don't correspond to common hardware capabilities.

True.

> I really don't think we want this in this form. For the Zephyr atomics API, we have toolchain support already on all targets (which is the ideal situation, absent known toolchain bugs or misfeatures). If you want to add a riscv-specific atomics API, that's obviously fine.

That is what I have actually done, the new "atomic_arc.h", in this case, it is  `include/zephyr/arch/riscv/atomic_riscv.h`, which will be activated only for RISC-V platform which sets CONFIG_ATOMIC_OPERATIONS_ARCH=y

andyross commented:
No, you're including it in a duplicate variant of the general API.  And it's AFAICT only available if you turn **off** the default one.

I'm saying this is a separate API.  Add a riscv-specific header, define it there, and don't tie it to the existing atomics layer.  And also add a test. :)

npitre commented:
The point is that, unless you have a compelling reason to do otherwise,
there is no advantage to not use the already available compiler 
built-ins which we don't have to maintain ourselves.

If you need extras not provided by the compiler built-ins then making a 
RISC-V specific implementation using a RISC-V specific name is fine.

Replacing the compiler built-ins should be done only if the built-ins 
are not OK, and that ought to be properly documented.

achech commented:
>  If you need extras not provided by the compiler built-ins then making a RISC-V specific implementation using a RISC-V specific name is fine.

 I'll do like that. Thanks.

fkokosinski commented on include/zephyr/arch/riscv/arch.h:
```diff
@@ -29,6 +29,9 @@
 #include <zephyr/arch/riscv/csr.h>
 #include <zephyr/arch/riscv/exception.h>
+#if defined(CONFIG_RISCV_ISA_EXT_A)
+#include <zephyr/arch/riscv/atomic.h>
```
This change should probably be moved somewhere here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/atomic.h#L30

fkokosinski reviewed (commented):
Hey, thanks for the PR! I left a couple of notes/questions

achech commented on tests/kernel/common/src/riscv_atomic.c:
```diff
@@ -0,0 +1,98 @@
+/*
```
Well, it is basically addition to atomic operations, I even thought of adding it in `/tests/kernel/common/src/atomic.c`
since test environment is the same.
I'll check where it can fit in `tests/arch/riscv`.

achech commented on include/zephyr/arch/riscv/atomic.h:
```diff
@@ -0,0 +1,142 @@
+					     atomic_val_t newval)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("lr.w.aq %0, %3\n"
+			"bne %0, %1, 1f\n"
+			"sc.d.aqrl %0, %2, %3\n"
+			"1:\n"
+			: "=&r"(ret)
+			: "r"(oldval), "r"(newval), "A"(*target)
+			: "memory");
+
+	return ret == 0;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_max(atomic_t *target, atomic_val_t value)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amomax.w.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(value), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_min(atomic_t *target, atomic_val_t value)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amomin.w.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(value), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+#endif /* CONFIG_64BIT */
```
Good point.

achech commented:
Hi @fkokosinski,
First I pushed the patch to check if the tests pass.

andyross commented on include/zephyr/arch/riscv/atomic.h:
```diff
@@ -0,0 +1,142 @@
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_lrsc_cas(atomic_t *target, atomic_val_t oldval,
+					     atomic_val_t newval)
```
What's the advantage of this vs. the compiler emitted one, which AFAICT looks the same.  An atomic_cas() in the current tree emits this for qemu_riscv32, for example, which looks broadly identical:

```
        return __atomic_compare_exchange_n(target, &old_value, new_value,
80000760:       07b00693                li      a3,123
80000764:       01878793                addi    a5,a5,24
80000768:       0f50000f                fence   iorw,ow
8000076c:       1407a52f                lr.w.aq a0,(a5)
80000770:       00e51563                bne     a0,a4,8000077a <main+0x42>
80000774:       1cd7a62f                sc.w.aq a2,a3,(a5)
80000778:       fa75                    bnez    a2,8000076c <main+0x34>
```

achech replied:
Right, removed it.

andyross reviewed (commented):
Remove -1, this is cleanly separated from the portable arch API now.  One question about the variant cas operation.

achech committed:
include: riscv: add RISC-V arch atomic instructions

Add RISC-V architecture specific atomic instructions, not present in
atomic_builtin.h

Signed-off-by: Aleksandar Cecaric <aleksandar.cecaric@nextsilicon.com>

achech commented:
Hi @andyross,
Can we have it this way ?

andyross reviewed (approved):
Sorry, this fell off the radar.  API looks good to me.
